lib/grpc-lwt/client.ml

```diff
@@ -78,14 +71,13 @@ module Rpc = struct
     [ `write ] H2.Body.t -> [ `read ] H2.Body.t Lwt.t -> 'a Lwt.t
 
   let bidirectional_streaming ~f write_body read_body =
-    let decoder_stream, decoder_push = Lwt_stream.create () in
-    Lwt.async (fun () ->
-        let+ read_body in
-        Connection.grpc_recv_streaming read_body decoder_push);
     let encoder_stream, encoder_push = Lwt_stream.create () in
-    Lwt.async (fun () ->
-        Connection.grpc_send_streaming_client write_body encoder_stream);
-    f encoder_push decoder_stream
+    let decoder_stream, decoder_push = Lwt_stream.create () in
+    let res = f encoder_push decoder_stream in
+    let* () = Connection.grpc_send_streaming_client write_body encoder_stream in
+    let* read_body = read_body in
+    Connection.grpc_recv_streaming read_body decoder_push;
+    res
```

Comment:
This preserves the semantics. The streaming usecase requires sending and receiving to happen in parallel, but this change makes it sequential.

Would this work with cancellation?
```suggestion
    let* () =
      Lwt.join
        [
          Connection.grpc_send_streaming_client write_body encoder_stream;
          (let+ read_body = read_body in
           Connection.grpc_recv_streaming read_body decoder_push);
        ]
    in
    res
```

Reply:
Yeah, that might be true, but then only for the `client_streaming` function, because both the `server_streaming` and `unary` functions atomically write a string to the encoder stream and then close it, so there is no concurrent scheduling possible anyway. I guess we really need test cases to keep this working. I will try the `join` version.

Reply:
Oh, `bidirectional_streaming` is an exported function. Ok, then I agree, it should keep its semantics.

Reply:
@ansiwen If I understand correctly, what @quernd meant that sending and receiving *have to* happen in parallel (by definition of bidi streaming). If you want to change the semantics of bidi streaming in a breaking way, feel free to (very few people use this lib) but the parallel sending/receiving bit is a *must*.

Reply:
@wokalski sure, as I wrote before, I didn't realise that it is an exported function. Still, we would need tests for that, because I can't test that with my use-case here.

Reply:
I updated the code with the concurrent version, and cancelation still works perfectly.

Reply:
I tested the updated version (well, not exhaustively, we're lacking proper tests, just whether our code still works) and it works well for bidirectional streaming 👍 
FWIW, right now all four types of gRPC (unary, server streaming, client streaming and bidirectional streaming) are currently implemented in terms of bidirectional streaming. It's a nice abstraction but it may not be the best solution. in particular, for unary calls, you could get rid of the `Lwt_stream` overhead by changing the implementation. It might pay off in your usecase from what I've seen.